packages/combo-box/src/vaadin-combo-box-mixin.js

```diff
@@ -576,17 +576,25 @@ export const ComboBoxMixin = (subclass) =>
 
     /** @private */
     _onEnter(e) {
-      // should close on enter when custom values are allowed, input field is cleared, or when an existing
-      // item is focused with keyboard. If auto open is disabled, under the same conditions, commit value.
-      if (
-        (this.opened || this.autoOpenDisabled) &&
-        (this.allowCustomValue || this._inputElementValue === '' || this._focusedIndex > -1)
-      ) {
-        this._closeOrCommit();
-
+      // do not commit value when custom values are disallowed and input value is not a valid option
+      // also stop propagation of the event, otherwise the user could submit a form while the input
+      // still contains an invalid value
+      if (!this.allowCustomValue && this._inputElementValue !== '' && this._focusedIndex < 0) {
         // Do not submit the surrounding form.
         e.preventDefault();
+        // Do not trigger global listeners
+        e.stopPropagation();
+        return;
+      }
+
+      const wasOpened = this.opened;
+      this._closeOrCommit();
 
+      // stop propagation of the enter event only if the dropdown was closed, this
+      // "consumes" the enter event for the action of closing the dropdown
+      if (wasOpened) {
+        // Do not submit the surrounding form.
+        e.preventDefault();
         // Do not trigger global listeners
         e.stopPropagation();
       }
```

Comment:
I wonder if `wasOpened` is really necessary here. The `if` statement could be moved `this._closeOrCommit()` instead:

```js
if (!this.allowCustomValue && this._inputElementValue !== '' && this._focusedIndex < 0) {
  e.preventDefault();
  e.stopPropagation();
  return;
}

if (this.opened) {
  e.preventDefault();
  e.stopPropagation();
}

this._closeOrCommit();
```

Reply:
Good idea, done.